### ISSUE TITLE
ModelCheckpoint doesn't count batches correctly when using `steps_per_execution`>1

### ISSUE THREAD

**Issue by jasnyj** [keras-team/keras#17589]
**System information**.
- Have I written custom code: yes
- OS Platform and Distribution: Arch Linux (Linux 6.1.11, x86_64)
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.11.0
- Python version: 3.10.9
- GPU model and memory: None, using CPU only, same issue with GPU on Colab (issue not related to GPU in any way)

**Describe the problem**.

When setting the parameter `steps_per_execution` to a value other than the default `None` in `model.compile(...)` and setting the `save_freq` parameter of the `ModelCheckpoint` callback to an integer to save the model after this many batches, the `ModelCheckpoint` callback doesn't count the number of batches trained on correctly: the very first batch (of the entire training) is not counted when using a `steps_per_execution` value greater than 1 and the callback therefore saves the model `steps_per_execution` batches too late.

**Standalone code to reproduce the issue**.

```
import tensorflow as tf
from tensorflow import keras

model = keras.Sequential([
    keras.layers.Dense(10, input_shape=(10,))
])

model.compile(
    loss="mse",
    steps_per_execution=16
)

batch_size = 32
num_examples = 1024
num_batches = num_examples // batch_size

x = tf.random.normal((num_examples, 10))
y = tf.random.normal((num_examples, 10))

callbacks = [
    keras.callbacks.ModelCheckpoint(
        "model.h5",
        verbose=1,
        save_freq=num_batches
    )
]

model.fit(x, y, batch_size=batch_size, callbacks=callbacks)
```

**Describe the current behavior**.

When using `steps_per_execution=None`, the callback correctly saves the model after the last batch of training and a line "Epoch 1: saving model to model.h5" is displayed.

However, when setting `steps_per_execution` to any value greater than 1, for instance `steps_per_execution=16`, the callback doesn't save the model after the last training batch of the training.

Note that in the standalone code example, the number of training batches do is divisible by `steps_per_execution` and therefore the `ModelCheckpoint` callback do is called at the end of the last training batch of the training, at which point in time `save_freq` batches have been trained on and the callback should thus save the model.

**Describe the expected behavior**.

Whatever value the parameter `steps_per_execution` is set to, the `ModelCheckpoint` callback saves the model after the last batch of the training (in the specific case of the standalone example code given above).

**Contributing**.

- Do you want to contribute a PR?: yes
- Briefly describe your candidate solution:

For now, in the `ModelCheckpoint` callback's constructor the variable `_last_batch_seen` is set to 0. 
When using `steps_per_execution=None`, on the very first batch of the training, the callback's `on_train_batch_end` method gets called with a `batch` argument equal to 0, making the `if` condition in the  `_should_save_on_batch` method true:

```
if batch <= self._last_batch_seen:  # New epoch.
    add_batches = batch + 1  # batches are zero-indexed.
```

When using `steps_per_execution=N`, the `on_train_batch_end` method only get called after the very first `N` batches of the training with its `batch` parameter set to `N-1`. The `else` path is therefore taken (because `_last_batch_seen=0`), which only counts `N-1` trained batches instead of `N`:

```
else:
    add_batches = batch - self._last_batch_seen
```

I propose to set `_last_batch_seen` to `None` in the `ModelCheckpoint`'s contructor and check for this condition:

```
if self._last_batch_seen == None or batch <= self._last_batch_seen:  # New epoch.
            add_batches = batch + 1  # batches are zero-indexed.
```

What I don't like about this solution is that it introduces an additional condition checked on each call of the callback. A second solution would be to set `_last_batch_seen` to a very high value in the constructor but which ? Does Python have some kind of INT_MAX constant ?

**Comment by tilakrayal** [comment]
@jasnyj,
I tried to execute the code with the alternative approach by providing steps_per_execution=1, N , 2, 3, 15 and 16 and for 1 and N it was executed as expected and provided the output.
![Screenshot 2023-02-23 6 52 46 PM](https://user-images.githubusercontent.com/81610181/220919198-b67e3fde-5f5f-4eaa-bd34-b4e892d7b6a5.png)

Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/f9788a88058354f07513ecf2b332beef/untitled990.ipynb). Also steps_per_execution expects Int. Defaults to 1. The number of batches to run during each [tf.function](https://www.tensorflow.org/api_docs/python/tf/function) call. Running multiple batches inside a single [tf.function](https://www.tensorflow.org/api_docs/python/tf/function) call can greatly improve performance on TPUs or small models with a large Python overhead. At most, one full epoch will be run each execution. If a number larger than the size of the epoch is passed, the execution will be truncated to the size of the epoch. Note that if steps_per_execution is set to N, Callback.on_batch_begin and Callback.on_batch_end methods will only be called every N batches (i.e. before/after each [tf.function](https://www.tensorflow.org/api_docs/python/tf/function) execution) and please take a look at the attached [reference](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/ModelCheckpoint). Thank you!

**Comment by jasnyj** [comment]
Please take a careful look back at your Colab's cells outputs. The line "Epoch 1: saving model to model.h5" is *not* displayed when using a `steps_per_execution` value equal to 16, 2, 3 or 15 which clearly shows the model is not being saved while it should be.

When I say it should be saved, I'm taking into account that "if steps_per_execution is set to N, Callback.on_batch_begin and Callback.on_batch_end methods will only be called every N batches (i.e. before/after each [tf.function](https://www.tensorflow.org/api_docs/python/tf/function) execution)".

Thanks.

**Comment by tilakrayal** [comment]
@jasnyj,
 If the Model is compiled with steps_per_execution=N, then the saving criteria will be checked every Nth batch. Note that if the saving isn't aligned to epochs, the monitored metric may potentially be less reliable (it could reflect as little as 1 batch, since the metrics get reset every epoch). Defaults to 'epoch'. Thank you!

**Comment by tilakrayal** [comment]
Closing this as stale. Please reopen if this is still a valid request. Thank you!

**Comment by jasnyj** [comment]
Fixed in PR #17632.
